Switch from bisection method to regula falsi method in precise models

stinebuu opened the pull request:
This PR fixes #1272 and fixes #403

The PR
- Replace the `bisectioning_` method with the [regula falsi](https://en.wikipedia.org/wiki/Regula_falsi) method (Illinois algorithm) in `iaf_psc_alpha_ps`, `iaf_psc_exp_ps`, `iaf_psc_exp_ps_lossless`.
- Fix membrane potential function for `iaf_psc_alpha_ps` so we are using the one for `alpha`.
- Switch to use the already defined `propagator_{31, 32}` where they were not already used.
- Use `std::exp()*x` instead `expm1()*x + x` throughout said models.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
 
 double
-nest::iaf_psc_alpha_ps::bisectioning_( const double dt ) const
+nest::iaf_psc_alpha_ps::V_m_root_function_( double t_step ) const
```
The function returns the difference of V_m and threshold, so I would call it `threshold_distance_`.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
+}
+
+double
+nest::iaf_psc_alpha_ps::regula_falsi_method_( const double dt ) const
```
Just `regular_falsi`.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
+  double V_m_func_b_k = V_m_root_function_( b_k );
 
-  double V_m_root = V_.V_m_before_;
+  assert( V_m_func_a_k * V_m_func_b_k <= 0 );
```
Maybe better to throw an exception is this fails. There could always be some strange corner case.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
+  assert( V_m_func_a_k * V_m_func_b_k <= 0 );
 
-  double div = 2.0;
+  int max_iter = 500;
```
Should be a constant, with capital letters. Also the 1e-14 in the termination criterion.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
-      root -= dt / div;
-    }
-    else
+    root = ( a_k * V_m_func_b_k - b_k * V_m_func_a_k ) / ( V_m_func_b_k - V_m_func_a_k );
```
Do we need to protect against a zero denominator?

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
-    }
-    else
+    root = ( a_k * V_m_func_b_k - b_k * V_m_func_a_k ) / ( V_m_func_b_k - V_m_func_a_k );
+    V_m_root = V_m_root_function_( root );
```
This is not V_m, but the difference from the threshold.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
+    root = ( a_k * V_m_func_b_k - b_k * V_m_func_a_k ) / ( V_m_func_b_k - V_m_func_a_k );
+    V_m_root = V_m_root_function_( root );
+
+    if ( std::abs( V_m_root ) <= 1e-14 )
```
I would avoid `<=` if possible, a simple `<` is cleaner and has the same effect here. The limit value should be defined as named constant.

heplesser commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -578,36 +575,78 @@ nest::iaf_psc_alpha_ps::emit_instant_spike_( Time const& origin, const long lag,
+  double root;
+  double V_m_root;
+
+  int side = 0;
```
I don't understand the logic behind "side" here and below, especially why it starts as zero and then will be -1 or +1 one forever as soon as it has taken on a different value than 0.  Could you explain this better? Possibly use an enumeration to have meaningful names instead of -1/0/+1?

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -253,13 +253,20 @@ class iaf_psc_alpha_ps : public Archiving_Node
+   * @param   double time step
+   * @returns difference between updated membrane potential and threshold
+   */
+  double V_m_root_function_( double t_step ) const;
```
Name should express what the function returns.

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -253,13 +253,20 @@ class iaf_psc_alpha_ps : public Archiving_Node
+  double V_m_root_function_( double t_step ) const;
+
+  /**
+   * Localize threshold crossing by using Illinois algorithm of regula falsi method.
```
Maybe add a reference to the algorithm, it may not be very well known.

heplesser commented on precise/iaf_psc_exp_ps.cpp:
```diff
@@ -540,36 +540,75 @@ nest::iaf_psc_exp_ps::emit_instant_spike_( const Time& origin, const long lag, c
+}
+
+double
+nest::iaf_psc_exp_ps::regula_falsi_method_( const double dt ) const
```
Instead of implementing this method for models that need it, could we implement it once i libnestutil? We would need to pass `V_m_root_function_` (with a better name) as a pointer-to-member-function, I think.

heplesser commented on testsuite/unittests/test_iaf_ps_dc_accuracy.sli:
```diff
@@ -340,8 +340,6 @@ V <- ( mV.) <- endl
 endl
 
 (               h in ms    ) <-
-(        alpha_canon [mV]) <-
```
Why do you remove these lines?

stinebuu replied:
`iaf_alpha_canon` was previously removed from the test as it is deprecated. I noticed the output was still here in the test when working on this PR because the test failed. I can add it again and remove it in another PR?

heplesser replied:
No, fixing it here is fine.

stinebuu commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -361,6 +358,7 @@ nest::iaf_psc_alpha_ps::update( Time const& origin, const long from, const long
       // update membrane potential
       if ( not S_.is_refractory_ )
       {
+        // If we do not use V_.expm1_tau_m_*S_.V_m_ + S_.V_m_ here, the accuracy is a bit worse.
```
@heplesser This is a very long answer to your question, but the accuracy compared to a theoretical value is higher when using `exp()`. If we use `S_.V_m_ * std::exp( -V_.h_ms_ / P_.tau_m_ )` instead of `V_.expm1_tau_m_ * S_.V_m_ + S_.V_m_` two tests fail, `testsuite/unittests/test_iaf_ps_dc_t_accuracy.sli` and `testsuite/unittests/test_iaf_ps_dc_accuracy.sli` (the two tests I have changed below). The tests "compares spikes generated to the theoretical value for different computation step sizes and interpolation orders". `testsuite/unittests/test_iaf_ps_dc_t_accuracy.sli` writes out information, and if we use `exp()` instead of `expm1()` we get, for interpolation order 3
```
order: 3, exact value of membrane potential after 5 ms is 4.70003629245736 ms.

               h in ms              alpha_presc [ms]      error           [ms]          delta_canon [ms]        error         [ms]
--------------------------------------------------------------------------------------------------------------------------------
                       0          4.70003629245736      8.88178419700125e-16          4.70003629245736      8.88178419700125e-16  
                      -1          4.70003629245736                         0          4.70003629245736                         0  
                      -2          4.70003629245736                         0          4.70003629245736                         0  
                      -3          4.70003629245735      6.21724893790088e-15          4.70003629245736      8.88178419700125e-16  
                      -4          4.70003629245737      1.33226762955019e-14          4.70003629245735      4.44089209850063e-15  
                      -5          4.70003629245736      1.77635683940025e-15          4.70003629245736      8.88178419700125e-16  
                      -6          4.70003629245733      2.48689957516035e-14          4.70003629245735      1.77635683940025e-15  
                      -7          4.70003629245728      7.72715225139109e-14          4.70003629245736      5.32907051820075e-15  
                      -8          4.70003629245749      1.35891298214119e-13          4.70003629245736      8.88178419700125e-16  
                      -9          4.70003629245741      5.24025267623074e-14          4.70003629245736                         0  
                     -10          4.70003629245698      3.77475828372553e-13          4.70003629245736      5.32907051820075e-15  
                     -11          4.70003629245835       9.9475983006414e-13          4.70003629245736      2.66453525910038e-15  
                     -12          4.70003629245631      1.04805053524615e-12          4.70003629245736      5.32907051820075e-15  
                     -13          4.70003629246496      7.60724816473157e-12          4.70003629245733      2.48689957516035e-14  
                     -14          4.70003629246858      1.12256870465899e-11          4.70003629245738       2.1316282072803e-14 
```

When we use `expm1()` as we do now however, we get
```
order: 3, exact value of membrane potential after 5 ms is 4.70003629245736 ms.

               h in ms              alpha_presc [ms]      error           [ms]          delta_canon [ms]        error         [ms]
--------------------------------------------------------------------------------------------------------------------------------
                       0          4.70003629245736      8.88178419700125e-16          4.70003629245736      8.88178419700125e-16  
                      -1          4.70003629245736      8.88178419700125e-16          4.70003629245736                         0  
                      -2          4.70003629245736      8.88178419700125e-16          4.70003629245736                         0  
                      -3          4.70003629245735      1.77635683940025e-15          4.70003629245736      8.88178419700125e-16  
                      -4          4.70003629245735      2.66453525910038e-15          4.70003629245735      4.44089209850063e-15  
                      -5          4.70003629245735      1.77635683940025e-15          4.70003629245736      8.88178419700125e-16  
                      -6          4.70003629245735      2.66453525910038e-15          4.70003629245735      1.77635683940025e-15  
                      -7          4.70003629245736      6.21724893790088e-15          4.70003629245736      5.32907051820075e-15  
                      -8          4.70003629245736      8.88178419700125e-16          4.70003629245736      8.88178419700125e-16  
                      -9          4.70003629245736                         0          4.70003629245736                         0  
                     -10          4.70003629245736      5.32907051820075e-15          4.70003629245736      5.32907051820075e-15  
                     -11          4.70003629245736       3.5527136788005e-15          4.70003629245736      2.66453525910038e-15  
                     -12          4.70003629245736        7.105427357601e-15          4.70003629245736      5.32907051820075e-15  
                     -13          4.70003629245733      2.39808173319034e-14          4.70003629245733      2.48689957516035e-14  
                     -14          4.70003629245738      2.04281036531029e-14          4.70003629245738       2.1316282072803e-14  
```
So for lower `h` the error compared to a theoretical value is lower when using `expm1()`. I can still use `exp()`, but then I will need to change the tolerance in the two tests to make them pass.

heplesser replied:
@stinebuu Thanks for the details! I think we need to take a look at some aspects of that test. The model names in the table header seem wrong (could we take them directly from the list of models tested instead of hardcoding them?), the `iaf_psc_[alpha,delta]_ps` models have no parameter "interpolation order" any more (so no reason to loop over orders), and `iaf_psc_exp_ps` and `iaf_psc_exp_lossless` are not tested.

This leaves the curious matter that the presumably more accurate numerical expression leads to less accurate results. It would be nice to understand that, but that would be a follow-up issue. For now, let us stick to the  V expm1 + V, which gives best performance. That then also should be used in the function computing V during smaller steps and during interpolation. In the comment, maybe point explicitly to this test.

Silmathoron commented:
Hi there, I'm sorry but I won't have time to review this in the foreseeable future.
Furthermore, I only implemented the previous version because nobody else could do it; I'm not used to working with propagators so I would have to redo all the math to check it, which is definitely not happening right now.
I would also like to remind you that I was explicitely asked to change the spike detection from interpolation to bisectioning in the original PR #1169 (see [here](https://github.com/nest/nest-simulator/pull/1169#issuecomment-499239564)), so I do not know why the change in method is now considered necessary, but I hope you made sure that this is indeed what is needed ^^

heplesser commented:
@Silmathoron Thanks for letting me know, I have assigned another reviewer. I had a look at the discussion in #1169 and I fear we got rather confused in that review (my fault and apologies!). Hanuschkin et al (2010) actually write in their paper that they skip use of bisection as suggested in earlier work by Brette, and just use Newton-Raphson. The "lossless" paper by Krishnan et al (2018) only deals with finding an interval and points out that one then needs to do some root-finding within that interval, but do not discuss any specific algorithm.

Experiments by @stinebuu in connection with this PR indicate that interpolation (regula falsi) requires about 10 times fewer steps than bisection.

heplesser commented on libnestutil/regula_falsi.h:
```diff
@@ -0,0 +1,112 @@
+
+      if ( last_threshold_sign == 1 )
+      {
+        // If threshold_dist_a_k and threshold_dist_root had the same sign in the last time step, we half the value of
```
I'd suggest to make the comment lines shorter to improve readibility, I think they should stay within the 80 char limit.

heplesser commented on nestkernel/node.h:
```diff
@@ -830,6 +830,13 @@ class Node
   virtual index get_local_device_id() const;
 
+  /**
+   * Difference between threshold and membrane potential for given time step.
```
Maybe add a note that this is used by precise neurons.

heplesser commented on precise/iaf_psc_alpha_ps.h:
```diff
@@ -172,6 +172,13 @@ class iaf_psc_alpha_ps : public Archiving_Node
+   * Difference between threshold and membrane potential for given time step.
+   * @param   double time step
+   * @returns difference between updated membrane potential and threshold
+   */
```
No need to repeat doxygen comment when overriding a virtual method.

stinebuu replied:
I have removed the virtual method.

heplesser commented on libnestutil/regula_falsi.h:
```diff
@@ -0,0 +1,112 @@
+ */
+template < typename CN >
+double
+regula_falsi( CN* node, double ( Node::*threshold_distance )( double ) const, const double dt )
```
@stinebuu Nice, and apologies for having complicated things more than necessary, as I had forgotten about the details of pointer to member function (too much Python programming lately ...). In contrast to Python, C++ does not have bound member functions, so we need to point the node anyways, and then we could just pass the node:

```c++
template < typename CN >
double 
regula_falsi( const CN& node, double dt )
...
```

heplesser commented on nestkernel/node.h:
```diff
@@ -830,6 +830,13 @@ class Node
+   * @param   double time step
+   * @returns difference between updated membrane potential and threshold
+   */
+  virtual double threshold_distance( double t_step ) const;
```
Actually, I wonder if we need `threshold_distance` as a virtual function in `Node` at all: Since `regula_falsi` has the node class as a template argument, it should work with all node classes providing `threshold_distance`, and for all others, it does not matter.

stinebuu replied:
When doing the pointer to member this seemed necessary, but I will try now with just pointing to the node!

stinebuu commented:
@heplesser I have removed interpolation from `test_iaf_ps_dc_t_accuracy.sli` and changed the output to be based on the neurons used. If I add `iaf_psc_exp_ps` and `iaf_psc_exp_ps_lossless` to the test, it fails and again it is because of the difference in accuracy between `exp()*x` and `expm1()*x + x`. Do you want me to add the neurons to the tests, and switch back to using `expm1`? Or I can lower the tolerance, or do you want me to see if I can figure out why we have the change in accuracy?

heplesser commented on libnestutil/regula_falsi.h:
```diff
@@ -30,16 +30,16 @@ namespace nest
 template < typename CN >
 double
-regula_falsi( CN* node, double ( Node::*threshold_distance )( double ) const, const double dt )
+regula_falsi( CN* node, const double dt )
```
I think `const CN& node` and a corresponding `regula_falsi(*this)` would be better.

stinebuu replied:
Done! :)

heplesser commented on libnestutil/regula_falsi.h:
```diff
@@ -39,7 +39,7 @@ namespace nest
 template < typename CN >
 double
-regula_falsi( CN* node, const double dt )
+regula_falsi( CN& node, const double dt )
```
Can't this be a `const` reference? I don't think we change the `node` in any way?

stinebuu replied:
Sorry, I was a bit quick and forgot the `const`.

heplesser commented:
@clinssen Could you take a look in the near future?

clinssen commented on precise/iaf_psc_exp_ps.h:
```diff
@@ -164,6 +164,13 @@ class iaf_psc_exp_ps : public Archiving_Node
   void set_status( const DictionaryDatum& );
 
+  /**
+   * Difference between threshold and membrane potential for given time step.
```
```suggestion
   * Based on the current state, compute the value of the membrane potential after taking a timestep of length
   * ``t_step``, and use it to compute the signed distance to spike threshold at that time. The internal state is not
   * actually updated (method is defined const).
```

clinssen commented on precise/iaf_psc_alpha_ps.cpp:
```diff
@@ -500,37 +503,32 @@ nest::iaf_psc_alpha_ps::handle( DataLoggingRequest& e )
 
     S_.V_m_ = ps_P30 * ( P_.I_e_ + S_.y_input_ ) + ps_P31_ex * S_.dI_ex_ + ps_P32_ex * S_.I_ex_ + ps_P31_in * S_.dI_in_
-      + ps_P32_in * S_.I_in_ + ps_e_Tau * S_.V_m_ + S_.V_m_;
+      + ps_P32_in * S_.I_in_ + S_.V_m_ * std::exp( -dt / P_.tau_m_ );
```
Can we rewrite this expression to its equivalent expm1 form for better floating-point precision?

heplesser replied:
We have this equation essentially in three places: lines 364f, 516f and 588ff. We should use the same implementation everywhere. In the `test_iaf_psc_dc...accurary.sli` tests, the `V * expm1() + V` curiously leads to smaller errors than `V * exp()` in the presence of pure DC input; so it may be the better choice.

heplesser commented:
@clinssen 
> * What happens to "canon" models?

There is only `iaf_psc_alpha_canon` left, for historical reasons. #1493 will follow that up.

> * Original `test_iaf_ps_dc_t_accuracy.sli` tests for different interpolation orders, why not any longer?

When we first introduced precise models around 2007, the time of threshold crossing was estimated by linear, quadratic or cubic interpolation and the user could select interpolation order as a model parameter.  We switched to iterative methods for threshold-crossing detection quite some time ago, eliminated the interpolation order parameter and never properly updated the tests.

heplesser commented on testsuite/unittests/test_iaf_ps_dc_t_accuracy.sli:
```diff
@@ -133,125 +131,78 @@ params begin userdict begin
+  endl endl
+  (The precise spike time is ) <- t <- ( ms.) <- endl endl
+
+  (                  h [ms]) <-
```
You are right! The table header here should really read `log_2 h` and for simplicities sake I would then just ignore the fact that `h` has units; writing the entirely correct `log_2(h / 1ms)` just seems too confusing. And this is test output ...

clinssen commented:
Thanks for the clarifications! In this case, additionally, please remove ``enum interpOrder`` from ``iaf_psc_alpha_ps.h`` and the comments block in front of it.

stinebuu commented on libnestutil/regula_falsi.h:
```diff
@@ -0,0 +1,113 @@
+    {
+      throw NumericalInstability( "regula_falsi: Regula falsi method did not converge" );
+    }
+  }
```
Good catch, we would have returned the `root` for the last iteration. I have added a `NumericalInstability`. There are now two right after each other, but I couldn't think of a better way of doing it.

heplesser replied:
@stinebu Doesn't the for-loop terminate when `iter == MAX_ITER`? And wouldn't it make the code even clearer to move the `return root;` directly into the `if` on line 72 above instead of `break`, and then just throw the exception here at the end unconditionally?

clinssen commented:
There are still some mentions of "bisectioning" in ``iaf_psc_exp_ps.h``.
